### PR TITLE
Fix over-indexing of lists and restore string indexing

### PR DESCRIPTION
**Pull request by aparajit-pratap**
### Purpose

JIRA: 
https://jira.autodesk.com/browse/DYN-995
https://jira.autodesk.com/browse/DYN-1056

image

image

image

image

### Declarations

Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [x] All tests pass using the self-service CI.
- [x] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions), and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

@mjkkirschner

### CODE REVIEW

File: src/NodeServices/Attributes.cs [hunk 1]
```diff
@@ -342,4 +342,26 @@ public class KeepReferenceThisAttribute : Attribute
...
+    /// to do method resolution and match argument(s) to the function parameter(s). 
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Method)]
+    public class AllowArrayPromotionAttribute : Attribute
+    {
+        public bool IsAllowed { get; private set; }
+
+        public AllowArrayPromotionAttribute()
+        {
+            IsAllowed = true;
```
Comment (mjkkirschner) [inline src/NodeServices/Attributes.cs]: so it looks like the default here was false previously? And now it's true? Am I interpreting that correctly? If so - does that have the potential for lots of behavior changes?

File: src/Libraries/DesignScriptBuiltin/Builtin.cs [hunk 2]
```diff
@@ -29,6 +30,17 @@ public static object ValueAtIndex(IList list, int index)
                 }
                 return list[index];
             }
+
+            public static object ValueAtIndex(string stringList, int index)
+            {
+                while (index < 0)
```
Comment (mjkkirschner) [inline src/Libraries/DesignScriptBuiltin/Builtin.cs]: what is the purpose of this loop?

File: src/NodeServices/Attributes.cs (hunk 1 above)
Comment (aparajit-pratap) [reply src/NodeServices/Attributes.cs]: Yup or should I add an attribute called `DisallowArrayPromotion` in which case we can get rid of the extra boolean flag.

File: src/Libraries/DesignScriptBuiltin/Builtin.cs (hunk 2 above)
Comment (aparajit-pratap) [reply src/Libraries/DesignScriptBuiltin/Builtin.cs]: Indexing from the end of the list with negative indices.

File: test/Engine/ProtoTest/Associative/MicroFeatureTests.cs [hunk 3]
```diff
@@ -1265,39 +1263,6 @@ public void TestDictionaryRegressMAGN619()
             thisTest.Verify("r", 5);
         }
 
-        [Test, Category("Failure")]
```
Comment (aparajit-pratap) [inline test/Engine/ProtoTest/Associative/MicroFeatureTests.cs]: This test doesn't make any sense.

File: src/Engine/ProtoCore/RuntimeStatus.cs [hunk 4]
```diff
@@ -18,7 +18,7 @@ public enum WarningID
             Default,
             AccessViolation,
             AmbiguousMethodDispatch,
-            AurgumentIsNotExpected,
+            ArgumentIsNotExpected,
```
Comment (mjkkirschner) [inline src/Engine/ProtoCore/RuntimeStatus.cs]: @aparajit-pratap I don't think we can change these now, they are public. 😢 maybe just add a todo?

**Comment by mjkkirschner** [comment]
@aparajit-pratap looks good except I don't think we can fix the enum names now.

**Comment by aparajit-pratap** [comment]
@aparajit-pratap these are changes in ProtoCore. I'm pretty sure no users use any public stuff from ProtoCore. Or are you saying this will cause binary incompatibility issues?

File: src/Engine/ProtoCore/Parser/AssociativeAST.cs [hunk 5]
```diff
@@ -1663,6 +1663,7 @@ public IEnumerable<string> ReturnKeys
         public string ObsoleteMessage { get; protected set; }
         public bool IsObsolete { get { return !string.IsNullOrEmpty(ObsoleteMessage); } }
         public bool IsLacingDisabled { get; protected set; }
+        public bool AllowArrayPromotion { get; protected set; } = true;
```
Comment (mjkkirschner) [inline src/Engine/ProtoCore/Parser/AssociativeAST.cs]: @aparajit-pratap should this also be added to the imperative AST methodAttributes class - does that exist?

**Comment by mjkkirschner** [comment]
@aparajit-pratap looks good to me - one question about imperative behavior vs associative?

File: src/Engine/ProtoCore/Parser/AssociativeAST.cs (hunk 5 above)
Comment (mjkkirschner) [reply src/Engine/ProtoCore/Parser/AssociativeAST.cs]: 😆

File: src/Engine/ProtoCore/FFI/CLRFFIFunctionPointer.cs [hunk 6]
```diff
@@ -5,8 +5,11 @@
 using ProtoCore.DSASM;
 using ProtoCore.Utils;
 using Autodesk.DesignScript.Runtime;
+using DesignScript.Builtin;
 using ProtoCore.Properties;
 using ProtoCore.Exceptions;
+using IndexOutOfRangeException = DesignScript.Builtin.IndexOutOfRangeException;
+using KeyNotFoundException = DesignScript.Builtin.KeyNotFoundException;
```
Comment (QilongTang) [inline src/Engine/ProtoCore/FFI/CLRFFIFunctionPointer.cs]: Can you make sure the usings are sorted?

File: src/Engine/ProtoCore/FFI/CLRObjectMarshaler.cs [hunk 7]
```diff
@@ -1332,7 +1332,7 @@ void core_Dispose(ProtoCore.RuntimeCore sender)
         {
             CLRObjectMarshaler marshaller = null;
             if (!mObjectMarshlers.TryGetValue(sender, out marshaller))
-                throw new KeyNotFoundException();
+                throw new System.Collections.Generic.KeyNotFoundException();
```
Comment (QilongTang) [inline src/Engine/ProtoCore/FFI/CLRObjectMarshaler.cs]: hmm This file has `System.Collections.Generic` using already I believe

File: src/Libraries/DesignScriptBuiltin/Builtin.cs [hunk 8]
```diff
@@ -16,9 +18,17 @@ public static class Get
...
             public static object ValueAtIndex(Dictionary dictionary, string key)
             {
-                return dictionary.ValueAtKey(key);
+                try
+                {
+                    return dictionary.ValueAtKey(key);
+                }
+                catch (System.Collections.Generic.KeyNotFoundException e)
+                {
+                    throw new KeyNotFoundException(e.Message);
```
Comment (QilongTang) [inline src/Libraries/DesignScriptBuiltin/Builtin.cs]: Same here, I wonder if it is some Resharp bug..

File: src/Engine/ProtoCore/FFI/CLRObjectMarshaler.cs (hunk 7 above)
Comment (mjkkirschner) [reply src/Engine/ProtoCore/FFI/CLRObjectMarshaler.cs]: this is to be explicit because we also have a key not found exception.

Comment (QilongTang) [reply src/Engine/ProtoCore/FFI/CLRObjectMarshaler.cs]: @mjkkirschner Then should we rename our own exception in a different task?

Comment (mjkkirschner) [reply src/Engine/ProtoCore/FFI/CLRObjectMarshaler.cs]: please a different task. our is 
`DesignScript.Builtin.KeyNotFoundException` 
@aparajit-pratap is aliasing them above.

**Pull request merged**